fastly/block_fastly_service_v1_director.go

```diff
@@ -0,0 +1,237 @@
+				ValidateFunc: validateDirectorType(),
+			},
+			"retries": {
+				Type:        schema.TypeInt,
+				Optional:    true,
+				Default:     5,
+				Description: "How many backends to search if it fails",
+			},
+		},
+	},
+}
+
+func processDirector(d *schema.ResourceData, conn *gofastly.Client, latestVersion int) error {
+	od, nd := d.GetChange("director")
+	if od == nil {
+		od = new(schema.Set)
+	}
+	if nd == nil {
+		nd = new(schema.Set)
+	}
+
+	ods := od.(*schema.Set)
+	nds := nd.(*schema.Set)
+
+	removeDirector := ods.Difference(nds).List()
+	addDirector := nds.Difference(ods).List()
+
+	// DELETE old director configurations
+	for _, dRaw := range removeDirector {
+		df := dRaw.(map[string]interface{})
+		opts := gofastly.DeleteDirectorInput{
+			Service: d.Id(),
+			Version: latestVersion,
+			Name:    df["name"].(string),
+		}
+
+		log.Printf("[DEBUG] Director Removal opts: %#v", opts)
+		err := conn.DeleteDirector(&opts)
+		if errRes, ok := err.(*gofastly.HTTPError); ok {
+			if errRes.StatusCode != 404 {
+				return err
+			}
+		} else if err != nil {
+			return err
+		}
+	}
+
+	// POST new/updated Director
+	for _, dRaw := range addDirector {
+		df := dRaw.(map[string]interface{})
+		opts := gofastly.CreateDirectorInput{
+			Service:  d.Id(),
+			Version:  latestVersion,
+			Name:     df["name"].(string),
+			Comment:  df["comment"].(string),
+			Shield:   df["shield"].(string),
+			Capacity: uint(df["capacity"].(int)),
+			Quorum:   uint(df["quorum"].(int)),
+			Retries:  uint(df["retries"].(int)),
+		}
+
+		switch df["type"].(int) {
+		case 1:
+			opts.Type = gofastly.DirectorTypeRandom
+		case 2:
+			opts.Type = gofastly.DirectorTypeRoundRobin
+		case 3:
+			opts.Type = gofastly.DirectorTypeHash
+		case 4:
+			opts.Type = gofastly.DirectorTypeClient
+		}
+
+		log.Printf("[DEBUG] Director Create opts: %#v", opts)
+		_, err := conn.CreateDirector(&opts)
+		if err != nil {
+			return err
+		}
+
+		if v, ok := df["backends"]; ok {
+			if len(v.(*schema.Set).List()) > 0 {
+				for _, b := range v.(*schema.Set).List() {
+					opts := gofastly.CreateDirectorBackendInput{
+						Service:  d.Id(),
+						Version:  latestVersion,
+						Director: df["name"].(string),
+						Backend:  b.(string),
+					}
+
+					log.Printf("[DEBUG] Director Backend Create opts: %#v", opts)
+					_, err := conn.CreateDirectorBackend(&opts)
+					if err != nil {
+						return err
+					}
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func readDirector(conn *gofastly.Client, d *schema.ResourceData, s *gofastly.ServiceDetail) error {
+	log.Printf("[DEBUG] Refreshing Directors for (%s)", d.Id())
+	directorList, err := conn.ListDirectors(&gofastly.ListDirectorsInput{
+		Service: d.Id(),
+		Version: s.ActiveVersion.Number,
+	})
+
+	if err != nil {
+		return fmt.Errorf("[ERR] Error looking up Directors for (%s), version (%v): %s", d.Id(), s.ActiveVersion.Number, err)
+	}
+
+	log.Printf("[DEBUG] Refreshing Backends for (%s)", d.Id())
```

Comment:
I like that you removed passing in the list of backends and keep everything contained in this function.

commit - https://github.com/opencredo/terraform-provider-fastly/pull/43/commits/c4a6c3c41be56a8348443ee1d950073b8e1fa9eb

Reply:
Thanks